### PR TITLE
[Flow] Add Redis Component

### LINKED ISSUE #2306: Flow: add redis_exporter
Create `prometheus.integration.redis_exporter` component

### PR DESCRIPTION
**Pull request by spartan0x117**
#### PR Description
Adds a flow component for `redis_exporter`.

Closes #2306 

#### PR Checklist

- [X] CHANGELOG updated
- [X] Documentation added
- [X] Tests updated

### CODE REVIEW

File: component/prometheus/integration/redis/redis.go
Comment (rfratto) [inline component/prometheus/integration/redis/redis.go:48]: For sensitive Flow fields, use [rivertypes.Secret](https://github.com/grafana/agent/blob/b722ea7f91a780194036f607a1cc54fe1e577a20/pkg/flow/rivertypes/secret.go#L17) like this: https://github.com/grafana/agent/blob/b722ea7f91a780194036f607a1cc54fe1e577a20/component/common/config/types.go#L23

`rivertypes.Secret` ensures that secrets get scrubbed when using the UI, and prevents Flow configs from accidentally converting a secret value back to a normal string.

Comment (spartan0x117) [reply component/prometheus/integration/redis/redis.go]: Got it, going forward would it best to cast to a prometheus `config.Secret` (e.g. `config_util.Secret(c.RedisPassword)`) or add a new case to [ConvertInto](https://github.com/grafana/agent/blob/b722ea7f91a780194036f607a1cc54fe1e577a20/pkg/flow/rivertypes/secret.go#L33)?

Comment (rfratto) [reply component/prometheus/integration/redis/redis.go]: Casting when converting to the integration type is the standard way of handling it 👍

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:14]: The link isn't working for me in the preview.  Might need to include http?  https://github.com/oliver006/redis_exporter

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:29]:
```suggestion
`redis_addr`                  | `string`   | Address (host and port) of the Redis instance to connect to. This can be a local Redis instance (e.g., `localhost:6379`) or the address of a remote server. | | yes
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:31]:
```suggestion
`redis_password`              | `string`   | Password of the Redis instance. | | no
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:32]:
```suggestion
`redis_password_file`         | `string`   | Path of a file containing a password. If defined, it takes precedence over redis_password. | | no
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:33]:
```suggestion
`redis_password_map_file`     | `string`   | Path of a file containing a JSON object which maps Redis URIs to passwords (e.g., `{"redis://localhost:6379": "sample_password"}`). | | no
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:37]:
```suggestion
`check_keys`                  | `string`   | Comma-separated list of key-patterns to export value and length/size, searched for with SCAN. | | no
`check_key_groups`            | `string`   | Comma-separated list of Lua regular expressions (regex) for grouping keys. When unset, no key groups are made. | | no
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:38]:
```suggestion
`check_key_groups_batch_size` | `int`      | Check key or key groups batch size hint for the underlying SCAN. The option name reflects key groups for backwards compatibility, but this applies to both key and key groups. | `10000` | no
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:39]:
```suggestion
`max_distinct_key_groups`     | `int`      | The maximum number of distinct key groups with the most memory utilization to present as distinct metrics per database. The leftover key groups are aggregated in the 'overflow' bucket. | `100` | no
```

Comment (karengermond) [reply docs/sources/flow/reference/components/prometheus.integration.redis.md:39]: Use present tense.  Google style guide section: https://developers.google.com/style/tense.

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [reply docs/sources/flow/reference/components/prometheus.integration.redis.md:37]: Hyphenate compound modifiers before a noun. We use Google Developer Style Guide -- here's the section: https://developers.google.com/style/hyphens#compound-modifiers

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:40]:
```suggestion
`check_single_keys`           | `string`   | Comma-separated list of single keys to export value and length/size. | | no
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:41]:
```suggestion
`check_streams`               | `string`   | Comma-separated list of stream-patterns to export info about streams, groups, and consumers to search for with SCAN. | | no
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:42]:
```suggestion
`check_single_streams`        | `string`   | Comma-separated list of single streams to export info about streams, groups, and consumers. | | no
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:43]:
```suggestion
`count_keys`                  | `string`   | Comma-separated list of individual keys to export counts for. | | no
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:50]:
```suggestion
`is_tile38`                   | `bool`     | Whether to scrape Tile38-specific metrics. | | no
```

Comment (karengermond) [reply docs/sources/flow/reference/components/prometheus.integration.redis.md:50]: Hyphenate compound modifiers before nouns.

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:52]:
```suggestion
`export_client_port`          | `bool`     | Whether to include the client's port when exporting the client list. Note that including the port increases the cardinality of all Redis metrics. | | no
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:53]:
```suggestion
`redis_metrics_only`          | `bool`     | Whether to just export metrics or to also export go runtime metrics. | | no
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:54]:
```suggestion
`ping_on_connect`             | `bool`     | Whether to ping the Redis instance after connecting. | | no
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:55]:
```suggestion
`incl_system_metrics`         | `bool`     | Whether to include system metrics, for example, redis_total_system_memory_bytes. | | no
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:69]:
```suggestion
For example, `targets` can either be passed to a `prometheus.relabel`
component to rewrite the metrics' label set, or to a `prometheus.scrape`
component that collects the exposed metrics.
```

**Review by karengermond — APPROVED** [review]
LGTM, left some edits.

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (rfratto) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:20]:
```suggestion
prometheus.integration.redis "LABEL" {
}
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (spartan0x117) [reply docs/sources/flow/reference/components/prometheus.integration.redis.md:37]: Good to know - thanks!

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (rfratto) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:43]: It's a little weird to me how many of these fields are a comma-separated list. Should we change them to use a `[]string` instead? 

It would introduce more differences between the Flow component and the existing integration, but it might be a better user experience which is harder to get wrong in the config file. 

WDYT @spartan0x117 @captncraig?

Comment (spartan0x117) [reply docs/sources/flow/reference/components/prometheus.integration.redis.md:43]: It'd definitely be a better user experience, generally. It _could_ be a tiny bit more tedious to switch from using the exporter directly or using the agent in non-flow-mode -> using flow, as it would require converting these values for the new flow config (granted that could be done pretty quickly using a simple script).

Comment (spartan0x117) [reply docs/sources/flow/reference/components/prometheus.integration.redis.md]: If we wanted to do that, I just pushed a commit that changes the flow config to accept `[]string`s and converts them to comma-separated strings in `Convert()`. I also tested a few with a minimal local example and verified the data being sent.

If we decide not to go that route I'll just have to revert that last commit before merging.

File: CHANGELOG.md
Comment (rfratto) [inline CHANGELOG.md:48]: These are duplicate entries now, should the other ones be removed?

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (rfratto) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:31]:
```suggestion
`redis_password`              | `secret`       | Password of the Redis instance. | | no
```

File: docs/sources/flow/reference/components/prometheus.integration.redis.md
Comment (rfratto) [inline docs/sources/flow/reference/components/prometheus.integration.redis.md:27]: I have similar concerns here with the length of descriptions. If something needs to be longer than a single sentence, can the remainder of the information be moved outside of the table?

For an example, check out `otelcol.exporter.otlp`, specifically the queue block: https://grafana.com/docs/agent/latest/flow/reference/components/otelcol.exporter.otlp/#queue-block

**Review by rfratto — COMMENTED** [review]
LGTM after the last comments I just left, but I'm not sure if we should merge this prior to #3049 or after

**Review by rfratto — APPROVED** [review]
LGTM! @jcreixell should this wait to get merged until after #3049?

**Comment by jcreixell** [comment]
@rfratto let's merge it and I will take care of the rebase

**Pull request merged**
